### PR TITLE
Implement a basic Messaging Service

### PR DESCRIPTION
**Pull request by lalexgap**
# Description
Introduces a new message service responsible for sending messages between participants.

This introduces the concept of a "MessageService". The responsibility of this message service is to send messages to opponents and trigger the `incomingMessageHandler` when receiving a message.

Longer-term the `wallet` will accept a `MessageServiceFactory`  from the `app`. This allows the `wallet` to create a message service while supplying a callback that can push the message into the wallet.

The initial implementation is a `TestMessageService` used for the majority of the `with-peers` tests. (Ledger funding was skipped since we expect it to change drastically soon). 

## Changes 
1. New `MessageServiceInterface` and associated types.
2. New `TestMessageService`
3. Updated `with-peer` tests to use `TestMessageService`

## How Has This Been Tested? 

The `TestMessageService` is used in `with-peers` test.

Unit tests for the `TestMessageService` have been omitted since it's only used in tests.

---
### Code quality
- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [x] This change does not have an unduly wide scope
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary (can be 0)
- [x] I have added tests that prove my fix is effective or that my feature works, if necessary
### Project management
- [x] I have applied the [appropriate labels](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#b304e56734a74dfbb341b8b4b27b1c0c)
- [x] I have [linked to all relevant issues](https://help.zenhub.com/support/solutions/articles/43000010350-connecting-pull-requests-to-github-issues) (can be 0)
- [x] I have [linked to all dependent issues](https://help.zenhub.com/support/solutions/articles/43000010349-create-github-issue-dependencies) (can be 0)
- [x] I have assigned myself to this PR
- [x] I have chosen the appropriate [pipeline](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#d534e68e7edc46fe8a4cda61b2258c4e) on zenhub for the linked issue

### CODE REVIEW

**Comment by lalexgap** [comment]
@andrewgordstewart I'm hoping this is enough to unblock you on #3362?

File: packages/server-wallet/src/message-service/types.ts
Comment (andrewgordstewart) [inline packages/server-wallet/src/message-service/types.ts:21]: nit:
```suggestion
  incomingMessageHandler: MessageHandler
```

Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts]: Done in 858df91

File: packages/server-wallet/jest/with-peers-setup-teardown.ts
Comment (lalexgap) [inline packages/server-wallet/jest/with-peers-setup-teardown.ts:38]: TODO:  missing description

File: packages/server-wallet/src/__test-with-peers__/crash-tolerance/crash-tolerance.test.ts
Comment (lalexgap) [inline packages/server-wallet/src/__test-with-peers__/crash-tolerance/crash-tolerance.test.ts:100]: TODO: Lots of `resultA2.outbox.map(o => o.params))`. Could be a helper or `send` could accept `o`

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (lalexgap) [inline packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:23]: TODO: Leave comment specifying this is temporary until lf is refactored

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (lalexgap) [inline packages/server-wallet/src/__test-with-peers__/utils.ts:8]: @andrewgordstewart you've written a lot of test helpers, is there a nicer way of doing this?

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (lalexgap) [inline packages/server-wallet/src/__test-with-peers__/utils.ts:35]: TODO: check for duplicates when creating

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (andrewgordstewart) [reply packages/server-wallet/src/__test-with-peers__/utils.ts:8]: What don't you find nice about this?

Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/utils.ts:8]: The main issue is that you have to `await` `expectLatestStateToMatch` which could be easy to miss when writing a test. Probably not worth worrying too much about it.

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/utils.ts]: c8d8eba

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:25]: 64ea12c

File: packages/server-wallet/jest/with-peers-setup-teardown.ts
Comment (lalexgap) [reply packages/server-wallet/jest/with-peers-setup-teardown.ts]: fece21e

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (andrewgordstewart) [reply packages/server-wallet/src/__test-with-peers__/utils.ts:11]: I don't think there's any way around that, right?

File: packages/server-wallet/src/message-service/types.ts
Comment (NiloCK) [inline packages/server-wallet/src/message-service/types.ts]: Do we have stated standards for declaring interfaces vs. types?

File: packages/server-wallet/src/message-service/types.ts
Comment (NiloCK) [inline packages/server-wallet/src/message-service/types.ts:26]:
```suggestion
   * Sends out the messages to the specified recipients.
```

File: packages/server-wallet/src/__test-with-peers__/crash-tolerance/crash-tolerance.test.ts
Comment (NiloCK) [inline packages/server-wallet/src/__test-with-peers__/crash-tolerance/crash-tolerance.test.ts:101]: Clarifying intent: the various `messageService.send` calls are being inserted into existing tests in order to ensure that sending / receiving messages doesn't impact previous behavior?

Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/crash-tolerance/crash-tolerance.test.ts:101]: Yup.  The intent here was to create a new messaging service and use the `with-peers` tests to ensure that using the new messaging service didn't break anything.

File: packages/server-wallet/src/message-service/types.ts
Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts]: I don't think we do!

Generally, I tend to favour an interface when defining something like an API but that's mostly based on habit. 

From the [typescript handbook](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces)
```
For the most part, you can choose based on personal preference, and TypeScript will tell you if it needs something to be the other kind of declaration. If you would like a heuristic, use interface until you need to use features from type.
```
so it seems like we should prefer interfaces until they no longer suffice?

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/utils.ts:10]: I thought there might be a more elegant way using a custom jest matcher? Either way, I don't think it's worth the time.

File: packages/server-wallet/src/message-service/types.ts
Comment (andrewgordstewart) [reply packages/server-wallet/src/message-service/types.ts]: Can a `Class` implement a `type`?

Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts]: Beat me to it (I also tried it [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyglgWzAGwgFXNAvFA3gQwC4A7AVwQCMIAnAXwCh6IAPMAe2uCgGNl8BnfrEQoIAYT6CoI1AgjFgQ+ElQZIeelC1QwpCsjjco+KDgCM9WkA))

Comment (NiloCK) [reply packages/server-wallet/src/message-service/types.ts]: The differences between interfaces and types are small. The features from type that might apply in the above heuristic are in creating and extending from union types. Interfaces support declaration merging which can sometimes matter for user-augmentation of a library's exported types. Anyway, the usage here is not a problem at all, I just wanted to ask because I saw both types and an interface in the same file.

Thanks, Effective Typescript!

File: packages/server-wallet/src/message-service/types.ts
Comment (NiloCK) [inline packages/server-wallet/src/message-service/types.ts:10]:
```suggestion
  message: Message
```
Probably an improvement with respect to code re-use and type consistency.

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (NiloCK) [inline packages/server-wallet/src/__test-with-peers__/utils.ts:50]:
```suggestion
  const messageHandler: MessageHandler = async (message, me) => {
    const matching = wallets.find(w => w.participantId === message.recipient);

    if (!matching) {
      throw new Error(`Invalid to value ${message.recipient}`);
```
Refactor (to match other suggestion)

File: packages/server-wallet/src/message-service/test-message-service.ts
Comment (NiloCK) [inline packages/server-wallet/src/message-service/test-message-service.ts:18]:
```suggestion
      await this._receive(message, this);
```

Refactor (to match other suggestion)

**Review by NiloCK — APPROVED** [review]
Comments aside, I understand and approve this work. I'm still pretty green, and the change is not tiny, so it may be a good idea for another set of eyes to have a look.

File: packages/server-wallet/src/message-service/types.ts
Comment (andrewgordstewart) [inline packages/server-wallet/src/message-service/types.ts:11]: Why does `MessageHandler` require a message service? Do you plan to remove this in a future PR?

A `MessageService` should be able to:
1. allow the "owner" of the service to send messages to others
2. allow the "owner" to decide how to handle messages received

In production, each wallet has their own `MessageService` instance. Internally, it will listen for incoming messages. (eg. if it is on an HTTP server, it may listen to the `/inbox` endpoint. If it is an HTTP client, it may set up server sent events against a `/outbox` endpoint.) The `MessageHandler` allows the wallet to say "MessageService, when you receive a message, push it into my wallet."

Understandably, in our `with-peers` test suite, there needs to be some kind of a "broker" that has a handle on all of the `TestMessageService` implementations. In this PR, `TestMessageService` acts both as a broker _and_ the `MessageService` used by all of the peers. That has some undesirable side effects, like:
- if you `destroy()` the MessageService, it is destroyed for both peers

Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts:11]: > Why does `MessageHandler` require a message service? Do you plan to remove this in a future PR?

This is done to allow the `MessageHandler` to easily call `mesageService.send` if it wants to. Since the `MessageHandler` callback is constructed before the `MessageService` is constructed we don't have a reference to the `MessageService` in the `MessageHandler`. By passing it in it allow for easily constructing a `MessageHandler` that pushes messages into the wallet and calles `messageService.send` to respond.

> Understandably, in our with-peers test suite, there needs to be some kind of a "broker" that has a handle on all of the TestMessageService implementations. In this PR, TestMessageService acts both as a broker and the MessageService used by all of the peers. That has some undesirable side effects, like:

I think that's a fair point that is acting as both a broker and message service. It seemed the most straightforward to have the `TestMessageService` act as the broker instead of having a more complicated setup of two message services communicating.

> if you destroy() the MessageService, it is destroyed for both peers

I think that's not too big of a drawback? I think it's pretty apparent that destroying the message service would destroy it for both peers.

Comment (andrewgordstewart) [reply packages/server-wallet/src/message-service/types.ts:11]: You're declaring a `MessageHandler` type here that is not a test type. But, the type is suited to fit the `TestMessageService`, and not the production use case.

In production, a `MessageHandler`
1. will not have a handle on the peer's `MessageService`
2. does not need a handle on the "local `MessageService`

Right?

Comment (andrewgordstewart) [reply packages/server-wallet/src/message-service/types.ts:11]: > I think that's not too big of a drawback? I think it's pretty apparent that destroying the message service would destroy it for both peers.

Since the test setup shares a `MessageService`, yes, I agree it's apparent. But it's also not representational of the production use case, where each wallet has a distinct `MessageService` instance.

Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts:11]: > You're declaring a `MessageHandler` type here that is not a test type. But, the type is suited to fit the `TestMessageService`, and not the production use case.
> 
> In production, a `MessageHandler`
> 
> 1. will not have a handle on the peer's `MessageService`
> 2. does not need a handle on the "local `MessageService`
> 
> Right?

I think 2. is not true?

I'm under the assumption that a production message service might be setup roughly like this:

```
function handleMessage(message:Message){
     const response = this.wallet.pushMessage(message);
    await messageService.send(response);
}
const messageService = createMessageService(handleMessage);
```

However, to accomplish this the `handleMessage` needs a reference to the `messageService`. Since the function is being constructed before the message service is constructed, I've passed it in as an argument.

Comment (andrewgordstewart) [reply packages/server-wallet/src/message-service/types.ts:11]: I have a different (albeit fuzzy) model
```
// channel manager

function handleMessage(message:Message){
     return this.wallet.pushMessage(message);
}
const messageService = createMessageService(handleMessage);
wallet.on('outbox', message => messageService.send(message));
```

I think your approach is fine. Perhaps better, since the `wallet` does not have to emit events with your approach. What are the drawbacks of your approach?

In any case, the presence of the circular dependency indeed makes it confusing. Can you clarify that in the code?

Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts:11]: > I think your approach is fine. Perhaps better, since the `wallet` does not have to emit events with your approach. What are the drawbacks of your approach?

I think the main drawback is that it's kinda confusing?

  In our [discussion](https://www.notion.so/statechannels/State-Channels-a19d0ee8714e49e0b7b07e889e97cedd?p=ff9740bfa8f04d4981cf74774d944f33) we came to the conclusion that we wanted to avoid `EventEmitter` so I tried exploring solution without it. I think I still prefer avoiding the `EventEmitter` even with the confusion of the current solution.

> In any case, the presence of the circular dependency indeed makes it confusing. Can you clarify that in the code?

Yup! It also might be worthwhile writing a `readme.md` for the message service (especially since longer-term other people besides us might write a message service)

Comment (andrewgordstewart) [reply packages/server-wallet/src/message-service/types.ts:11]: An alternative is that the `MessageService` can be an event emitter that emits `'messageReceived'`. This allows the `ChannelManager` to attach a callback after the `MessageService` is instantiated.

File: packages/server-wallet/jest/with-peers-setup-teardown.ts
Comment (lalexgap) [reply packages/server-wallet/jest/with-peers-setup-teardown.ts:57]: That's more concise but slightly easier to shoot yourself in the foot. IE: One of these fails but it's not immediately clear which one: `crashAndRestart(['A'])`  or `crashAndRestart(['a'])`

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts:106]: One drawback of the messaging services is that it becomes slightly more difficult to check the results at every step (since the messages are automatically sent). Perhaps a `pause` method could make these checks easier? Although I'm not certain it's worth the effort?

Comment (andrewgordstewart) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts:106]: What's the goal of the test? Is it to assert that a specific sequence of messages happens in a specific order?

If that's the goal, then the `TestMessageService` can be used to that effect:
- It can _record_ which messages are exchanged, in what order.
- We can then _expect_ a statement about the messages that are exchanged

Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts:106]: I think the previous checks weren't looking at the `messages` but at the `channelResult` to check if the channel is in the correct state.

If we want to inspect the `channelResult` (or `message` payloads) at each step does it make more sense to just have the test manually call `pushMessage` like before? It seems like adding a `TestMessageService` then having to scan through the history is more work then just manually handing the pushMessage results.

File: packages/server-wallet/src/message-service/test-message-service.ts
Comment (lalexgap) [reply packages/server-wallet/src/message-service/test-message-service.ts]: Resolved in 0accce4

File: packages/server-wallet/src/__test-with-peers__/utils.ts
Comment (lalexgap) [reply packages/server-wallet/src/__test-with-peers__/utils.ts]: Resolved in 0accce4

File: packages/server-wallet/src/message-service/types.ts
Comment (lalexgap) [reply packages/server-wallet/src/message-service/types.ts]: Resolved in 0accce4

File: packages/server-wallet/jest/with-peers-setup-teardown.ts
Comment (andrewgordstewart) [reply packages/server-wallet/jest/with-peers-setup-teardown.ts:58]: If you wanted it to be safer, you could accept `('a' | 'b')[]`?

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts
Comment (andrewgordstewart) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/direct-funding.test.ts:106]: Let me restate: What _should_ the goal be of the `with-peers` tests?

- Should the goal be to ensure that the correct messages are passed, when one party calls `createChannels`?
- Should the goal be to ensure that when one party calls `createChannels`, that the promise resolves? and at the end, the party has the right number of funded channels?

> the previous checks weren't looking at the messages but at the channelResult to check if the channel is in the correct state.

I don't think it's necessarily wrong to have this duplicated, but there are already "unit tests" of these crankers that assert on the `channelResult`. So, maybe this test doesn't also need to do that.

> It seems like adding a TestMessageService then having to scan through the history is more work then just manually handing the pushMessage results.

IMO it's not a question about how much work it is, it's about what the test is testing. If it's meant to be testing what messages get passed back and forth, then I think writing the `TestMessageService` _once_ to record messages that are sent is worth it. You can then use the `TestMessageService` to assert that:
- in _this_ situation, these messages are exchanged
- in _that_ situation, those messages are exchanged

Again, I do not know if that's what we're trying to test here.

**Pull request merged**
